### ISSUE TITLE
plugin is unable to generate debugging class name

### ISSUE THREAD

**Issue by sylvain-hamel** [styled-components/babel-plugin-styled-components#108]
We have a HOC that calls to `styled`.  We use that HOC a lot and because our components don't use `styled` directly, it seems like the plug-in is not able to generate class names based on the component name.

This code _breaks_ the plug-in:

```
import * as styled1  from 'styled-components';
const styled = styled1.default;
const MySuperComponent = styled(({children, className}) =>
    (<div className={className}>{children}</div>))`color: red`;
```

I suppose that our HOC somehow creates the same conditions that the plug-in does not support.

Is this fixable? It there anything we can do in our HOC to give the plug-in it a hint?

**Comment by mxstbr** [comment]
Hmm, I think we could support this. Mind adding a failing test case based on the other tests? That'd be super helpful, thanks!

**Comment by philpl** [comment]
I think we can just target styled, css, injectGlobal, and keyframes regardless of where those variables are defined... Because otherwise the detection code would get ridiculous.

I think that'd be a fair assumption to make. Wdyt, @mxstbr?

**Comment by mxstbr** [comment]
> Because otherwise the detection code would get ridiculous.

I don't understand, what does the detection have to do anything? We already check that there's an import from `styled-components` and then we check if there's a `styled()` function call? It should only be a matter of adding `.withConfig` at the end of the styled function call no matter what's inside it, I don't think this has anything to do with detection.

**Comment by philpl** [comment]
@mxstbr I mean, code like @sylvain-hamel's breaks because he's presumably importing a local file instead of styled-components. So maybe we should get rid of the import/require-based detection.

**Comment by mxstbr** [comment]
Sorry? He's importing from `styled-components` though? I don't 100% understand what you're saying.

**Comment by mxstbr** [comment]
Oh you mean that this breaks the detection?

```JS
import * as styled1  from 'styled-components';
const styled = styled1.default;
```

Possibly. @sylvain-hamel can you try making that a normal `import styled from 'styled-components'` import?

**Comment by sylvain-hamel** [comment]
Hi all, I'll send a test case shortly.

**Comment by sylvain-hamel** [comment]
Are tests supposed to run on Windows? They passed on my Mac Book but 17 of them fail on my windows computer.

**Comment by sylvain-hamel** [comment]
Ok, so here is the code.

This is the HOC:

```
// turns `styled` into a composable hoc
const composeStyledComponent = (value, ...rest) => 
    BaseComponent => styled(BaseComponent)(value, ...rest);

```

Code that uses the HOC:
```
const enhance = composeStyledComponent`color:red`;
const Component = ({className}) => <div className={className}>hello</div>;
const EnhancedComponent = enhance(Component);
```

And then some JSX somewhere

```
<EnhancedComponent/>
```

**Comment by sylvain-hamel** [comment]
I'd like an update on this issue if possible. Is the code I provided clear? Can the plug-in be modified to support that use case? Thanks

**Comment by probablyup** [comment]
We do the hash calculation based on the location of the call to styled, so
that makes sense if all your styled calls are in the exact same file.

**Comment by Atsman** [comment]
@probablyup yeah I understood that. Any ideas how to fix that? Quite ofther i want to be able to use style interpolation like

```javascript
// button.js
import styled from 'styled-components'
import withStyled from '...'
import icon from '...'

const StyledButton = styled.button`
// ....
`

const Button = ({ text, ...props }) => (
  <StyledButton {...props}>
    <icon />
    { text }
  </StyledButton>
)

// and here is what i want to do, because later probably 
// i would need to add styles related to positioning. 
export default withStyled(Button)
```
```javascript
import styled from 'styled-components'
import button from 'components/button'

const StyledCard = styled.div`
  display: flex
  ${button} {
    margin-top: 10px;
  }
`

const Card = () => (
  <StyledCard>
   <Button />
  </StyledCard>
)
```

So what i see here is quite similar to all macro problems, now that thing doesn't compose, and the only way is to pass styles as arguments, or to create a wrapper, quite sad.

Any ideas?

**Comment by tabrindle** [comment]
Running into what I think is the same issue, but there's another distressing element to it - it works on macOS, and fails on Linux (CI). 

```javascript
import styled from 'styled-components';

export const styledSomething = Component => styled(Component)`
  display: flex;
`;
```

I didn't write any of this HoC code, and to be honest it seems unnecessary here, but why would it work on one platform and not the other? I'm curious about this one.
